ssl/ssl_key_share.cc

```diff
@@ -176,20 +178,550 @@ class X25519KeyShare : public SSLKeyShare {
+        if (!key_shares_[i] ||
+            !key_shares_[i]->Finish(&component_secret, out_alert, component_key) ||
+            !CBB_add_bytes(&hybrid_shared_secret, component_secret.data(), component_secret.size())) {
+          CBB_cleanup(&hybrid_shared_secret);
+          OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+          return false;
+        }
+
+        peer_key_index += component_key_size;
+      }
+
+      // Final validation that |peer_key| was the correct size
+      if (peer_key_index != peer_key.size()) {
+        CBB_cleanup(&hybrid_shared_secret);
+        *out_alert = SSL_AD_DECODE_ERROR;
+        OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+        return false;
+      }
+
+      // Retain the hybrid shared secret for our use.
+      if (!CBBFinishArray(&hybrid_shared_secret, out_secret)) {
+        CBB_cleanup(&hybrid_shared_secret);
+        OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+        return false;
+      }
+
+      // Success; clear the alert
+      *out_alert = 0;
+      return true;
+    }
+
+  private:
+   bool init_component_key_shares() {
+      // The component key shares should only be initialized once
+      if (hybrid_group_) {
+        return false;
+      }
+
+     for (const HybridGroup &hybrid_group : HybridGroups()) {
+       if (group_id_ == hybrid_group.group_id) {
+         hybrid_group_ = &hybrid_group;
+         for (size_t i = 0; i < NUM_HYBRID_COMPONENTS; i++) {
+           key_shares_[i] = SSLKeyShare::Create(hybrid_group.component_group_ids[i]);
+         }
+         return true;
+       }
+     }
+      return false;
+   }
+
+   uint16_t group_id_;
+   const HybridGroup *hybrid_group_;
+   UniquePtr<SSLKeyShare> key_shares_[NUM_HYBRID_COMPONENTS];
+};
+
 CONSTEXPR_ARRAY NamedGroup kNamedGroups[] = {
     {NID_secp224r1, SSL_GROUP_SECP224R1, "P-224", "secp224r1"},
     {NID_X9_62_prime256v1, SSL_GROUP_SECP256R1, "P-256", "prime256v1"},
     {NID_secp384r1, SSL_GROUP_SECP384R1, "P-384", "secp384r1"},
     {NID_secp521r1, SSL_GROUP_SECP521R1, "P-521", "secp521r1"},
     {NID_X25519, SSL_GROUP_X25519, "X25519", "x25519"},
+    {NID_SecP256r1Kyber768Draft00, SSL_GROUP_SECP256R1_KYBER768_DRAFT00, "SecP256r1Kyber768Draft00", ""},
+    {NID_X25519Kyber768Draft00, SSL_GROUP_X25519_KYBER768_DRAFT00, "X25519Kyber768Draft00", ""},
 };
 
-}  // namespace
+CONSTEXPR_ARRAY uint16_t kPQGroups[] = {
+    SSL_GROUP_KYBER512_R3,
+    SSL_GROUP_KYBER768_R3,
+    SSL_GROUP_KYBER1024_R3,
+    SSL_GROUP_SECP256R1_KYBER768_DRAFT00,
+    SSL_GROUP_X25519_KYBER768_DRAFT00
+};
+
+// The key share sizes are taken from the corresponding specification.
+//
+// Kyber Round 3: https://pq-crystals.org/kyber/data/kyber-specification-round3-20210804.pdf
+// SECP256R1:     https://datatracker.ietf.org/doc/html/rfc8446#section-4.2.8.2
+// X25519:        https://datatracker.ietf.org/doc/html/rfc8446#section-4.2.8.2
+static const size_t p256_share_size = ((32 * 2) + 1);
+static const size_t x25519_share_size = 32;
```

Comment:
Should the hybrid key share care about the size of it's sub-key shares? Or just let them use read/write as much as they need and let the sub-keyshare throw an error if the size is wrong?

Reply:
The ECKeyShare will fail with an error if the KeyShare data passed to it is too large, so the HybridKeyShare class needs to know where to split the two component KeyShares in it to avoid those errors.